test/cpp/jit/test_interpreter.cpp

```diff
@@ -20,88 +26,97 @@ graph(%a.1 : Tensor,
-      std::vector<IValue> stack({a, b});
-      interp.run(stack);
-      ASSERT_FALSE(stack[2].toBool());
-    }
-    {
-      auto a = at::zeros({2, 2}, at::kFloat);
-      auto b = at::ones({3, 3}, at::kFloat);
-      a.set_requires_grad(true);
-      a = a.to(at::kCUDA); // Device mismatch
-      std::vector<IValue> stack({a, b});
-      interp.run(stack);
-      ASSERT_FALSE(stack[2].toBool());
-    }
+    return InterpreterState(function);
   }
+};
 
-  try { // Test empty Typecheck raises an internal assertion
-    auto graph = std::make_shared<Graph>();
-    std::unordered_map<std::string, Value*> vmap;
-    parseIR(
-        R"IR(
-graph(%a.1 : Tensor,
-      %b.1 : Tensor):
-  %type_matched : bool = prim::TypeCheck()
-  return (%type_matched)
-  )IR",
-        &*graph,
-        vmap);
-  } catch (const std::exception& e) {
-  }
-  try { // Test for assertion if num_inputs + 1 != num_outputs
-    auto graph = std::make_shared<Graph>();
-    std::unordered_map<std::string, Value*> vmap;
-    parseIR(
-        R"IR(
-graph(%a.1 : Tensor,
-      %b.1 : Tensor):
-  %type_matched : bool = prim::TypeCheck(%a.1)
-  return (%type_matched)
-  )IR",
-        &*graph,
-        vmap);
-  } catch (const std::exception& e) {
-  }
+TEST_F(TypeCheckTest, MatchingType) {
+  // TypeCheck yields to true! Shape, grad and device matches.
+  auto a = at::zeros({2, 2}, at::kFloat);
+  auto b = at::ones({3, 3}, at::kFloat);
+  a.set_requires_grad(true);
+  a = a.to(at::kCPU);
+  std::vector<IValue> stack({a, b});
+  interp.run(stack);
+  ASSERT_TRUE(exactlyEqual(stack[0].toTensor(), a));
+  ASSERT_TRUE(exactlyEqual(stack[1].toTensor(), b));
+  ASSERT_TRUE(stack[2].toBool());
+}
+
+TEST_F(TypeCheckTest, SizeMismatch) {
+  auto a = at::zeros({2, 2}, at::kFloat);
+  auto b = at::ones({2, 2}, at::kFloat); // Size mismatch
+  a.set_requires_grad(true);
+  a = a.to(at::kCPU);
+  std::vector<IValue> stack({a, b});
+  interp.run(stack);
+  ASSERT_FALSE(stack[2].toBool());
 }
-void testInterp() {
+
+TEST_F(TypeCheckTest, GradientMismatch) {
+  auto a = at::zeros({2, 2}, at::kFloat);
+  auto b = at::ones({3, 3}, at::kFloat);
+  a = a.to(at::kCPU);
+  a.set_requires_grad(false); // Gradient mismatch
+  std::vector<IValue> stack({a, b});
+  interp.run(stack);
+  ASSERT_FALSE(stack[2].toBool());
+}
+
+TEST_F(TypeCheckTest, ScalarTypeMismatch) {
+  auto a = at::zeros({2, 2}, at::kFloat);
+  auto b = at::ones({3, 3}, at::kFloat);
+  a = a.to(at::kCPU);
+  a.set_requires_grad(true);
+  a = a.to(at::kInt); // Scalar type mismatch
+  std::vector<IValue> stack({a, b});
+  interp.run(stack);
+  ASSERT_FALSE(stack[2].toBool());
+}
+
+TEST_F(TypeCheckTest, DeviceMismatch_CUDA) {
+  auto a = at::zeros({2, 2}, at::kFloat);
+  auto b = at::ones({3, 3}, at::kFloat);
+  a.set_requires_grad(true);
+  a = a.to(at::kCUDA); // Device mismatch
+  std::vector<IValue> stack({a, b});
+  interp.run(stack);
+  ASSERT_FALSE(stack[2].toBool());
+}
+
+// TODO: These tests weren't doing anything.
```

Comment:
Do these tests pass if we enable them now?

Reply:
There were a few tests that weren't running, because they were defined in a test file but not registered in `tests.h`. If they are commented out, it's because they failed when I re-enabled them :(